src/inmanta/server/services/compilerservice.py

```diff
@@ -601,13 +602,104 @@ async def request_recompile(
         :return: the compile id of the requested compile and any warnings produced during the request
+        """
+        return await self._request_recompile(
+            env,
+            force_update,
+            do_export,
+            remote_id,
+            metadata,
+            env_vars,
+            partial,
+            removed_resource_sets,
+            exporter_plugin,
+            notify_failed_compile,
+            failed_compile_message,
+        )
+
+    async def request_recompile_in_transaction(
+        self,
+        connection: Connection,
+        env: data.Environment,
+        force_update: bool,
+        do_export: bool,
+        remote_id: uuid.UUID,
+        metadata: Optional[JsonType] = None,
+        env_vars: Optional[Mapping[str, str]] = None,
+        partial: bool = False,
+        removed_resource_sets: Optional[List[str]] = None,
+        exporter_plugin: Optional[str] = None,
+        notify_failed_compile: Optional[bool] = None,
+        failed_compile_message: Optional[str] = None,
+    ) -> Tuple[Optional[uuid.UUID], Warnings]:
+        """
+        This method is similar to the `request_recompile()` method, but should be used when the compile request
+        is part of an ongoing database transaction. The given `connection` attribute should be within the scope
+        of that database transaction. It is required to call `CompileService.notify_compile_request_committed()`
+        right after the transaction commits.
+        """
+        return await self._request_recompile(
+            env,
+            force_update,
+            do_export,
+            remote_id,
+            metadata,
+            env_vars,
+            partial,
+            removed_resource_sets,
+            exporter_plugin,
+            notify_failed_compile,
+            failed_compile_message,
+            connection=connection,
+            in_db_transaction=True,
+        )
+
+    async def notify_compile_request_committed(self, compile_id: uuid.UUID) -> None:
+        """
+        This method must be called when the `request_recompile_in_transaction()` method was called and the associated
+        transaction committed successfully. It must be called right after the transaction committed. If the transaction was
+        aborted, this method most not be called.
+
+        This method is safe with respect to server restarts. If the server restarts after the transaction was committed,
+        but before this method was invoked, the server will automatically recover from this and run the requested compile
+        without any need to call this method.
+        """
+        async with self._queue_count_cache_lock:
+            self._queue_count_cache += 1
+        compile_obj: data.Compile = await data.Compile.get_by_id(compile_id)
+        if compile_obj.started:
+            raise Exception(f"Compile with id {compile_obj.id} was already started at {compile_obj.started}.")
```

Comment:
I don't think we should raise an exception here. This can happen when another compile finished more or less simultaneously with the transaction commit. Then this method call and the start of the compile would race with each other. Since both have the same purpose that's a success scenario.